### ISSUE TITLE
Migration to VS2019 and latest WDK

### ISSUE THREAD

**Issue by billziss-gh** [billziss-gh/winfsp#228]
I am looking to migrate WinFsp to VS2019. The ideal scenario would be to support both VS2015 and VS2019 with latest WDK and the same solution/project files.

Unfortunately due to issues with the different versions of VS and WDK this goal does not seem easy to achieve. As an alternative I might migrate the whole project to VS2019 with latest WDK; before I do so I want to verify that this would not create problems for anyone.

**Comment by sganis** [comment]
No issues, go ahead.

**Comment by cyberhaus** [comment]
This would be great thanks.

**Comment by cyberhaus** [comment]
Since this an open source product, would be good to see that minimal VS2019 version would be community.

**Comment by dworkin** [comment]
Upgrading to VS2019 would be a good move.

**Comment by billziss-gh** [comment]
I have made a few attempts to modify the project files so that they work with both VS2015 and VS2019 without modifications. Unfortunately I have not been successful in eliminating all upgrade prompts. The promised "Visual Studio round-tripping" does not seem to work for me.

I have come to the decision to simply upgrade the project for VS2019 and leave other VS versions behind. However I have to wait until AppVeyor adds VS2019 and latest WDK in their images, because WinFsp uses AppVeyor for a lot of its automated testing.

**Comment by billziss-gh** [comment]
I am now able to load the projects without the upgrade dialog appearing by changing the `PlatformToolset` element as follows:

```
<PlatformToolset>$(DefaultPlatformToolset)</PlatformToolset>
```

This uses the `v140` toolset on VS2015 and the `v142` toolset on VS2019. It appears that this is enough for VS2019 to not initiate a project upgrade.

However there are a number of other issues with VS2019, some of them breaking:

- **BREAKING**: `LINK : fatal error LNK1295: '/DEBUG:fastlink' not compatible with '/PDBSTRIPPED' specification; link without '/PDBSTRIPPED'`
    - This is breaking because WinFsp uses `/PDBSTRIPPED` to produce public symbols that are included in the installer. It also uses `/DEBUG` to produce the PDB files; unfortunately Microsoft in their infinite wisdom changed the meaning of `/DEBUG` to mean `/DEBUG:fastlink` in VS2017.
    - For more information see here: https://developercommunity.visualstudio.com/content/problem/162958/fastlink-wrongly-used-by-vs-2017-linker.html

- `warning MSB8038: Spectre mitigation is enabled but Spectre mitigated libraries are not found.  Verify that the Visual Studio Workload includes the Spectre mitigated libraries.  See https://aka.ms/Ofhn4c for more information.`

I will update this comment as I find more information/workarounds.

**Comment by billziss-gh** [comment]
I believe that I have successfully managed to modify the project files in a way to make Visual Studio round-tripping possible, although it was more difficult than it should have been.

Commit 369da895d3bc409473d456fb551f6498ef194751 includes all changes:

- The `PlatformToolset` setting was changed to `$(DefaultPlatformToolset)` as explained above.
- The `GenerateDebugInformation` setting is set (according to the value of `$(DefaultPlatformToolset)`) to either `true` (for VS2015) or `DebugFull` (for VS2017+).
- Spectre mitigation is disabled for the time being until we assess the impact of having it enabled.

With these changes the WinFsp projects should be loadable in VS2015 - VS2019 without problems. Please report any issues if you find otherwise.

**Comment by sganis** [comment]
Your proposed changes work fine in vs2019.
In the dll project, WindowsTargetPlatformVersion could be changed to
$(LastestTargetPlatformVersion), instead of 8.1 as it is now.

**Comment by billziss-gh** [comment]
@sganis I was under the impression that the `LastestTargetPlatformVersion` Does not exist in VS2015? I will try it tomorrow.

Also see this link:
https://developercommunity.visualstudio.com/content/problem/140294/windowstargetplatformversion-makes-it-impossible-t.html

**Comment by dworkin** [comment]
The dotnet component targets .NET 3.5, which cannot be installed with VS2019 (I guess it can only be installed with older Visual Studio versions).  Leaving out that line from the project file will automatically target the lowest available framework version, which is probably what is desired.

Other projects target platform 8.1 which again is not offered by the VS2019 installer.  Falling back to the default targets the latest platform, rather than the earliest one that can be compiled for, so this might require some more conditional changes in version.properties?

**Comment by billziss-gh** [comment]
@dworkin good catches. I have both VS2015 and VS2019 on my system and did not catch those.

**Comment by billziss-gh** [comment]
With commit 195f3bf92d805d37f004d6df852cafa76cfec97b I have modified `WindowsTargetPlatformVersion` to be `LatestTargetPlatformVersion`. VS2015 displays this as `8.1` and VS2019 displays it as "10.0 (latest installed version)". Both build the WinFsp solution as well as the sample projects fine on my system.

Regarding the change to .NET 3.5: I am not certain we should change that. The reason is that WinFsp targets Windows 7, which shipped with .NET 3.5 [[reference](https://superuser.com/questions/109207/what-version-of-the-net-framework-is-included-in-windows-7build-7600)].

I have also found that the VS2019 installer includes a .NET Framework 3.5 component.

![Capture](https://user-images.githubusercontent.com/15330948/56823565-1c483780-6809-11e9-900c-f7caf3bf865c.PNG)
